### PR TITLE
FLIP

### PR DESCRIPTION
**Pull request by PhylomatX**
+ Added methods for implementing a FLIP simulation for liquids.
+ Added demo of FLIP simulation
+ Added masked extrapolation method for tensors and for CenteredGrid / StaggeredGrid
+ Added small tests for extrapolation method
+ Added position shift methods to geometries
+ Added runge-kutta advection scheme with velocity-dependent extrapolation
+ Rewrote distribute_points method
+ Changed duplicate_handling in PointCloud sampling method

### CODE REVIEW

File: phi/physics/flip.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        v_change_field = projected_v_field - initial_v_field
+        v_change_field, _ = field.extp_sgrid(v_change_field, smask, 1)  # conserves falling shapes (no hard_bcs here!)
+        v_change = v_change_field.sample_at(v_particle.elements.center)
+        return PointCloud(v_particle.elements, values=v_particle.values + v_change,
+                          add_overlapping=v_particle._add_overlapping)
+    else:
+        # PIC
+        v_div_free_field, _ = field.extp_sgrid(projected_v_field, smask, 1)
+        v_values = v_div_free_field.sample_at(v_particle.elements.center)
+        return PointCloud(v_particle.elements, values=v_values,
```
Comment (holl-) [inline phi/physics/flip.py:97]:
```suggestion
        return v_particle.with_(values=v_values)
```

File: phi/physics/flip.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    Returns:
+        Projected velocity field and corresponding pressure field
+    """
+    v_field, _ = field.extp_sgrid(v_field, smask, 1)  # extrapolation conserves falling shapes
+    v_field *= bcs  # Enforces boundary conditions after extrapolation
+    div = field.divergence(v_field) * cmask
+
+    def laplace(p):
+        # TODO: prefactor of pressure should not have any effect
```
Comment (holl-) [inline phi/physics/flip.py:62]: Does it though?

File: phi/math/_nd.py [hunk 3]
```diff
@@ -239,6 +239,48 @@ def shift(x: Tensor,
     return offset_tensors
 
 
+def masked_extp(x: Tensor, mask: Tensor, size: int = 1) -> Tuple[Tensor, Tensor]:
```
Comment (holl-) [inline phi/math/_nd.py:242]:
```suggestion
def extrapolate_valid(values: Tensor, valid: Tensor, distance_cells: int = 1) -> Tuple[Tensor, Tensor]:
```

File: phi/geom/_box.py [hunk 4]
```diff
@@ -97,6 +97,22 @@ def approximate_signed_distance(self, location):
         distance = math.abs(location - center) - extent * 0.5
         return math.max(distance, 'vector')
 
+    def shift_positions(self, positions: Tensor, outward: bool = True, shift_amount: float = 0) -> Tensor:
```
Comment (holl-) [inline phi/geom/_box.py:100]: This method needs documentation and the name is not descriptive enough.
How about `repel` or `push`?

Also it seems, this method does not return the new positions but the delta. Wouldn't most applications require the new positions?

File: phi/field/_grid.py [hunk 5]
```diff
@@ -320,3 +321,43 @@ def _validate_staggered_values(values: TensorStack):
...
+        size: Number of extrapolation steps
+
+    Returns:
+        CenteredGrid with extrapolated values and binary CenteredGrid marking all extrapolated positions.
+    """
+    new_tensor, new_mask = masked_extp(cgrid.values, mask.values, size)
+    return CenteredGrid(new_tensor, cgrid.box, cgrid.extrapolation), CenteredGrid(new_mask, mask.box, mask.extrapolation)
+
+
+def extp_sgrid(sgrid: StaggeredGrid, mask: StaggeredGrid, size: int = 1) -> Tuple[StaggeredGrid, StaggeredGrid]:
```
Comment (holl-) [inline phi/field/_grid.py:343]: See comment above.

**Review by holl- — CHANGES_REQUESTED** [review]
The code is mostly well-documented and well-structured.
However, some function names should be changed to make the code easier to understand.
Also, some functions can be merged into a single function, see attached comments.

**Comment by PhylomatX** [comment]
+ Applied all refactorings
+ Merged / Removed some of the methods
+ Implemented new merge method for PointClouds with & operator

**Pull request merged**